modules/um_callerid.c

```diff
@@ -0,0 +1,243 @@
+		return -1;
+	}
+
+	user_modes['G'] = find_umode_slot();
+	if (!user_modes['G'])
+	{
+		ierror("um_callerid: unable to allocate usermode slot for +G; unloading module.");
+		return -1;
+	}
+
+	construct_umodebuf();
+
+	add_isupport("CALLERID", isupport_string, "g");
+
+	return 0;
+}
+
+static void
+um_callerid_moddeinit(void)
+{
+	user_modes['g'] = 0;
+	user_modes['G'] = 0;
+	construct_umodebuf();
+
+	delete_isupport("CALLERID");
+}
+
+#define IsSetCallerID(c)	((c->umodes & user_modes['g']) == user_modes['g'])
+#define IsSetRelaxedCallerID(c)	((c->umodes & user_modes['G']) == user_modes['G'])
+
+static const char um_callerid_desc[] =
+	"Provides usermode +g which restricts messages from unauthorized users.";
+
+static bool
+has_common_channel(struct Client *source_p, struct Client *target_p)
+{
+	rb_dlink_node *ptr;
+
+	RB_DLINK_FOREACH(ptr, source_p->user->channel.head)
+	{
+		struct membership *msptr = ptr->data;
+		if (IsMember(target_p, msptr->chptr))
+			return msptr->chptr;
+	}
+
+	return NULL;
+}
+
+static bool
+allow_message(struct Client *source_p, struct Client *target_p)
+{
+	if (!MyClient(target_p))
+		return true;
+
+	if (!IsSetCallerID(target_p))
+		return true;
+
+	if (IsSetRelaxedCallerID(target_p) && has_common_channel(source_p, target_p))
+		return true;
+
+	if (IsServer(source_p))
+		return true;
+
+	/* XXX: controversial?  allow opers to send through +g */
+	if (IsOper(source_p))
+		return true;
+
+	if (accept_message(source_p, target_p))
+		return true;
+
+	return false;
+}
+
+static void
+send_callerid_notice(enum message_type msgtype, struct Client *source_p, struct Client *target_p)
+{
+	if (!MyClient(target_p))
+		return;
+
+	if (msgtype == MESSAGE_TYPE_NOTICE)
+		return;
+
+	sendto_one_numeric(source_p, ERR_TARGUMODEG, form_str(ERR_TARGUMODEG),
+		target_p->name);
+
+	if ((target_p->localClient->last_caller_id_time + ConfigFileEntry.caller_id_wait) < rb_current_time())
+	{
+		sendto_one_numeric(source_p, RPL_TARGNOTIFY, form_str(RPL_TARGNOTIFY),
+			target_p->name);
+
+		sendto_one(target_p, form_str(RPL_UMODEGMSG),
+			   me.name, target_p->name, source_p->name,
+			   source_p->username, source_p->host);
+
+		target_p->localClient->last_caller_id_time = rb_current_time();
+	}
+}
+
+static bool
+add_callerid_accept_for_source(enum message_type msgtype, struct Client *source_p, struct Client *target_p)
+{
+	/*
+	 * XXX: Controversial? Allow target users to send replies
+	 * through a +g.  Rationale is that people can presently use +g
+	 * as a way to taunt users, e.g. harass them and hide behind +g
+	 * as a way of griefing.  --nenolod
+	 */
+	if(msgtype != MESSAGE_TYPE_NOTICE &&
+			IsSetCallerID(source_p) &&
+			!accept_message(target_p, source_p) &&
+			!IsOper(target_p))
+	{
+		if(rb_dlink_list_length(&source_p->localClient->allow_list) <
+				(unsigned long)ConfigFileEntry.max_accept)
+		{
+			rb_dlinkAddAlloc(target_p, &source_p->localClient->allow_list);
```

Comment:
This will crash on a multi-server network; it is being called on the wrong server.